### PR TITLE
[GEOT-7030] Missing YSLD support for rule vendor options

### PR DESCRIPTION
**Pull request by sikeoka**
[GEOT-7030](https://osgeo-org.atlassian.net/browse/GEOT-7030) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

This PR updates the YSLD style format to handle vendor options for rules.

  

# Checklist

- [X] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [X] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [X] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [X] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [X] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [X] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [X] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [X] Bug fixes and small new features are presented as a single commit.
- [X] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Comment by aaime** [comment]
@jodygarnett this one is up your alley I think.
@sikeoka lots of test failures:

```
2021-11-30T22:37:20.2920205Z [ERROR] Errors: 
2021-11-30T22:37:20.2923170Z [ERROR]   SldTransformerTest.testPointSimple:53->transform:2039 » Emitter expected NodeE...
2021-11-30T22:37:20.2931934Z [ERROR]   SldTransformerTest.testPointWithAttribute:337->transform:2039 » Emitter expect...
2021-11-30T22:37:20.2934009Z [ERROR]   SldTransformerTest.testPointWithGraphic:130->transform:2039 » Emitter expected...
2021-11-30T22:37:20.2935745Z [ERROR]   SldTransformerTest.testPointWithLabel:461->transform:2039 » Emitter expected N...
2021-11-30T22:37:20.2937398Z [ERROR]   SldTransformerTest.testPointWithLegend:176->transform:2039 » Emitter expected ...
2021-11-30T22:37:20.2939212Z [ERROR]   SldTransformerTest.testPointWithRotatedLabel:593->transform:2039 » Emitter exp...
2021-11-30T22:37:20.2941021Z [ERROR]   SldTransformerTest.testPointWithRotation:373->transform:2039 » Emitter expecte...
2021-11-30T22:37:20.2942881Z [ERROR]   SldTransformerTest.testPointWithRuleVendorOption:640->transform:2039 » Emitter
2021-11-30T22:37:20.2948827Z [ERROR]   SldTransformerTest.testPointWithScale:246->transform:2039 » Emitter expected N...
2021-11-30T22:37:20.2954948Z [ERROR]   SldTransformerTest.testPointWithStroke:90->transform:2039 » Emitter expected N...
2021-11-30T22:37:20.2956778Z [ERROR]   SldTransformerTest.testPointWithStyledLabel:521->transform:2039 » Emitter expe...
2021-11-30T22:37:20.2958800Z [ERROR]   SldTransformerTest.testPointWithTransparentTriangle:412->transform:2039 » Emitter
2021-11-30T22:37:20.2959957Z [INFO] 
```

**Comment by sikeoka** [comment]
I had a mistake in the original commit but it's fixed now.

**Comment by sikeoka** [comment]
If there's no one to review this module can this just be merged if everything passes?

**Comment by aaime** [comment]
If there is no one to review we should consider calling the module unmaintained and pushing it down to unsupported land (or find a new maintainer). @jodygarnett I thought you had an interest in the module?

**Comment by jodygarnett** [comment]
Excellent, I was wanting see this improvement made.

File: modules/extension/ysld/src/test/java/org/geotools/ysld/encode/YsldEncodeCookbookTest.java [hunk 1]
```diff
@@ -619,6 +619,46 @@ public void testPointWithRotatedLabel() throws Exception {
...
+        //              <OnlineResource xlink:type="simple" xlink:href="smileyface.png" />
+        //             <Format>image/png</Format>
+        //           </ExternalGraphic>
+        //           <Size>20</Size>
+        //         </Graphic>
+        //       </PointSymbolizer>
+        //       <VendorOption name="inclusion">legendOnly</VendorOption>
+        //     </Rule>
+        //   </FeatureTypeStyle>
+        // </UserStyle>
```
Comment (jodygarnett) [inline modules/extension/ysld/src/test/java/org/geotools/ysld/encode/YsldEncodeCookbookTest.java:654]: Do we really need the commended out sld here?

File: modules/extension/ysld/src/test/java/org/geotools/ysld/parse/YsldParseCookbookTest.java [hunk 2]
```diff
@@ -604,6 +606,45 @@ public void testPointWithRotatedLabel() throws Exception {
...
+        //              <OnlineResource xlink:type="simple" xlink:href="smileyface.png" />
+        //             <Format>image/png</Format>
+        //           </ExternalGraphic>
+        //           <Size>20</Size>
+        //         </Graphic>
+        //       </PointSymbolizer>
+        //       <VendorOption name="inclusion">legendOnly</VendorOption>
+        //     </Rule>
+        //   </FeatureTypeStyle>
+        // </UserStyle>
```
Comment (jodygarnett) [inline modules/extension/ysld/src/test/java/org/geotools/ysld/parse/YsldParseCookbookTest.java:641]: The cookbook tests are here to ensure the examples from the documentation work as advertised, so one fix is to add this to the cookbook.

**Review by jodygarnett — CHANGES_REQUESTED** [review]
Thanks for the clean addition, to complete this I would like to complete the doc checklist for the ylsd page https://docs.geotools.org/latest/userguide/extension/ysld.html

**Comment by jodygarnett** [comment]
I added the documentaiton change, when the builds have completed I will rebase this PR. Thanks again for the contribution.

**Review by jodygarnett — APPROVED** [review]

**Comment by jodygarnett** [comment]
It is failing in docs:
```
/home/runner/work/geotools/geotools/docs/user/tutorial/datastore/strategy.rst:245:Could not lex literal_block as "java". Highlighting skipped.
```

Appears unrelated.

**Comment by aaime** [comment]
It is unrelated, see this other PR for an analysis and a solution (if you want to get green builds, you can try to rebase this one, or merge with main):
https://github.com/geotools/geotools/pull/3724#issuecomment-1003533144

**Comment by jodygarnett** [comment]
Thanks, I rebased with main locally have pushed up the changes after local testing.

**Comment by jodygarnett** [comment]
Thanks for the fix @sikeoka I have backported to 26.x, let me know if you need it backported to 25.x.

**Pull request closed without merging**